### PR TITLE
Add `providers lock` Command

### PR DESCRIPTION
**Pull request by philnielsen**
Adds feature requested in #201. This allows programatic generation of`.terraform.lock.hcl` files via this libary.

### CODE REVIEW

**Comment by philnielsen** [comment]
@radeksimko this is a follow up to #201 that i've finally gotten a chance to work on.

File: tfexec/providers_lock.go [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+func (tf *Terraform) ProvidersLock(ctx context.Context, opts ...ProvidersLockOption) error {
+	lockCmd := tf.providersLockCmd(ctx, opts...)
+
+	err := tf.runTerraformCmd(ctx, lockCmd)
+	if err != nil {
+		return err
+	}
+
+	return err
+}
```
Comment (radeksimko) [inline tfexec/providers_lock.go:47]: This command [was only introduced in 0.14](https://github.com/hashicorp/terraform/commit/e270291f194043486bc6c46759308ffa8bdec239), and so we should add some additional checks and return a human-readable error for older versions.

See how we perform similar check for `validate` here: https://github.com/hashicorp/terraform-exec/blob/43e54be55ebc3ce07d381660dcafcb45f567e67c/tfexec/validate.go#L15-L19

File: tfexec/options.go [hunk 2]
```diff
@@ -202,6 +229,15 @@ func PluginDir(pluginDir string) *PluginDirOption {
 	return &PluginDirOption{pluginDir}
 }
 
+type ProviderOption struct {
+	provider string
+}
+
+// Provider Represents the provider positional argument
```
Comment (radeksimko) [inline tfexec/options.go:236]:
```suggestion
// Provider represents the positional argument (provider source address)
```

File: tfexec/options.go [hunk 3]
```diff
@@ -178,6 +187,15 @@ func LockTimeout(lockTimeout string) *LockTimeoutOption {
 	return &LockTimeoutOption{lockTimeout}
 }
 
+type NetMirrorOption struct {
+	netMirror string
+}
+
+// NetMirror Represents the -fs-mirror option
```
Comment (radeksimko) [inline tfexec/options.go:194]:
```suggestion
// NetMirror represents the -net-mirror option (base URL of a network mirror)
```

File: tfexec/options.go [hunk 4]
```diff
@@ -117,6 +117,15 @@ func DryRun(dryRun bool) *DryRunOption {
 	return &DryRunOption{dryRun}
 }
 
+type FSMirrorOption struct {
+	fsMirror string
+}
+
+// FSMirror Represents the -fs-mirror option
```
Comment (radeksimko) [inline tfexec/options.go:124]:
```suggestion
// FSMirror represents the -fs-mirror option (path to filesystem mirror directory)
```

File: tfexec/providers_lock.go [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+
+func (opt *PlatformOption) configureProvidersLock(conf *providersLockConfig) {
+	conf.platforms = append(conf.platforms, opt.platform)
+}
+
+func (opt *ProviderOption) configureProvidersLock(conf *providersLockConfig) {
+	conf.providers = append(conf.providers, opt.provider)
+}
+
+// ProvidersLock represents the terraform providers lock
```
Comment (radeksimko) [inline tfexec/providers_lock.go:37]:
```suggestion
// ProvidersLock represents the `terraform providers lock` command
```

**Review by radeksimko — COMMENTED** [review]
Thank you for the PR.

Aside from my in-line comments it looks good!

**Review by radeksimko — APPROVED** [review]
👍🏻

**Pull request merged**